server/src/main/java/org/apache/iotdb/db/engine/migration/MigrationManager.java

```diff
@@ -0,0 +1,478 @@
+          .toString();
+
+  protected MigrationManager() {
+    init();
+  }
+
+  // singleton
+  private static class MigrationManagerHolder {
+    private MigrationManagerHolder() {}
+
+    private static final MigrationManager INSTANCE = new MigrationManager();
+  }
+
+  public static MigrationManager getInstance() {
+    return MigrationManagerHolder.INSTANCE;
+  }
+
+  public synchronized void init() {
+    if (initialized) {
+      return;
+    }
+
+    try {
+      logWriter = new MigrationLogWriter(LOG_FILE_NAME);
+    } catch (FileNotFoundException e) {
+      logger.error("Cannot find/create log for migration.");
+    }
+
+    // read from logReader
+    try {
+      MigrationLogReader logReader = new MigrationLogReader(LOG_FILE_NAME);
+      Set<Long> errorSet = new HashSet<>();
+
+      while (logReader.hasNext()) {
+        MigrationLog log = logReader.next();
+
+        switch (log.type) {
+          case SET:
+            setMigrationFromLog(
+                log.taskId,
+                log.storageGroup,
+                FSFactoryProducer.getFSFactory().getFile(log.targetDirPath),
+                log.ttl,
+                log.startTime);
+            break;
+          case UNSET:
+            unsetMigrationFromLog(log.taskId);
+            break;
+          case START:
+            // if task started but didn't finish, then error occurred
+            errorSet.add(log.taskId);
+            break;
+          case PAUSE:
+            errorSet.remove(log.taskId);
+            pauseMigrationFromLog(log.taskId);
+            break;
+          case UNPAUSE:
+            unpauseMigrationFromLog(log.taskId);
+            break;
+          case FINISHED:
+            // finished task => remove from list and remove from potential error task
+            errorSet.remove(log.taskId);
+            migrationTasks.remove(log.taskId);
+            finishFromLog(log.taskId);
+            break;
+          case ERROR:
+            // already put error in log
+            errorSet.remove(log.taskId);
+            errorFromLog(log.taskId);
+            break;
+          default:
+            logger.error("read migration log: unknown type");
+        }
+      }
+
+      // for each task in errorSet, the task started but didn't finish (an error)
+      for (long errIndex : errorSet) {
+        if (migrationTasks.containsKey(errIndex)) {
+          // write to log and set task in ERROR in memory
+          logWriter.error(migrationTasks.get(errIndex));
+          errorFromLog(errIndex);
+        } else {
+          logger.error("unknown error index");
+        }
+      }
+
+    } catch (IOException e) {
+      logger.error("Cannot read log for migration.");
+    }
+
+    // finished migrating the tsfiles that were in process
+    try {
+      MigratingFileLogManager.getInstance().recover();
+    } catch (IOException e) {
+      logger.error("migratingfile could not recover");
+    }
```

Comment:
You should recover before the system is ready, please reference the code `registerManager.register(StorageEngine.getInstance());` in the IoTDB class.